gh-114863: What's new in Python 3.13: JIT compiler (#114862)
python/cpython: Doc/whatsnew/3.13.rst

  over 70% of integer type checks in our own benchmarks.

* Loop splitting -- after the first iteration, we gain a lot more type
  information. Thus, we peel the first iteration of loops to produce
  an optimized body that exploits this additional type information.
  This also achieves a similar effect to an optimization called
  loop-invariant code motion, but only for guards.

* Globals to constant promotion -- global value loads become constant
  loads, speeding them up and also allowing for more constant propagation.

* This section is non-exhaustive and will be updated with further
  optimizations, up till CPython 3.13's release.

(Tier 2 optimizer contributed by Ken Jin, with implementation help
by Guido van Rossum, Mark Shannon, and Jules Poon. Special thanks
to Manuel Rigger and Martin Henz.)

Execution Engine
----------------

There are two execution engines for Tier 2 IR.

The first is the Tier 2 interpreter, but it is mostly intended for debugging
the earlier stages of the optimization pipeline. If the JIT is not
enabled, the Tier 2 interpreter can be invoked by passing Python the
``-X uops`` option or by setting the ``PYTHON_UOPS`` environment
variable to ``1``.

> The second is machine code. When the ``--enable-experimental-jit``

Reviewer: ```suggestion
The second is the JIT. When the ``--enable-experimental-jit``
```
("machine code" is not an execution engine)

build-time option is used, the optimized Tier 2 IR is translated to machine
code, which is then executed. This does not require additional
runtime options.

The machine code translation process uses a technique called
*copy-and-patch*. It has no runtime dependencies, but there is a new
build-time dependency on LLVM. The main benefit of this technique is

> rapid compilation times, reported as orders of magnitudes faster versus

Reviewer: ```suggestion
fast compilation, reported as orders of magnitudes faster versus
```
(or: small compilation times)

* Constant propagation -- through forward data-flow analysis, we can reduce
  expressions like ::

    a = 1
    b = 2
    c = a + b

 to ::

    a = 1
    b = 2
    c = 3

* Guard elimination -- through a combination of constant and type information,
  we can eliminate type checks and other guards associated with operations.
  These guards validate specialized operations, but add a slight bit of
  overhead. For example, integer addition needs a type check that checks
  both operands are integers. As a proof of concept, we managed to eliminate
  over 70% of integer type checks in our own benchmarks.

* Loop splitting -- after the first iteration, we gain a lot more type
  information. Thus, we peel the first iteration of loops to produce
  an optimized body that exploits this additional type information.
  This also achieves a similar effect to an optimization called
  loop-invariant code motion, but only for guards.

* Globals to constant promotion -- global value loads become constant
  loads, speeding them up and also allowing for more constant propagation.

* This section is non-exhaustive and will be updated with further

>   optimizations, up till CPython 3.13's release.

Reviewer 2: A suggestion, but just fine to leave it as is too:
```suggestion
  optimizations, up till CPython 3.13's beta release.
```

>   optimizations, up till CPython 3.13's beta release.

Reviewer 3: ```suggestion
  optimizations, until CPython 3.13's beta release.
```

(Tier 2 optimizer contributed by Ken Jin, with implementation help
by Guido van Rossum, Mark Shannon, and Jules Poon. Special thanks
to Manuel Rigger and Martin Henz.)

Execution Engine
----------------

> There are two execution engines for Tier 2 IR.

Reviewer 3: ```suggestion
There are two execution engines for Tier 2 IR: the Tier 2 interpreter and the Just-in-Time (JIT) compiler.
```

It's jarring to see the JIT mentioned in the paragraph below before it is introduced, so I think introducing them both here helps.

> The first is the Tier 2 interpreter, but it is mostly intended for debugging

Reviewer 3: ```suggestion
The Tier 2 interpreter is mostly intended for debugging
```

the earlier stages of the optimization pipeline. If the JIT is not
enabled, the Tier 2 interpreter can be invoked by passing Python the
``-X uops`` option or by setting the ``PYTHON_UOPS`` environment
variable to ``1``.

> The second is the JIT compiler. When the ``--enable-experimental-jit``

Reviewer 3: ```suggestion
The other execution engine is the JIT compiler. When the ``--enable-experimental-jit``
```

build-time option is used, the optimized Tier 2 IR is translated to machine
code, which is then executed. This does not require additional
runtime options.

The machine code translation process uses a technique called
*copy-and-patch*. It has no runtime dependencies, but there is a new

> build-time dependency on LLVM. The main benefit of this technique is

Reviewer 3: ```suggestion
build-time dependency on `LLVM <https://llvm.org>`_. The main benefit of this technique is
```

fast compilation, reported as orders of magnitudes faster versus
traditional compilation techniques in the paper linked below. The code
produced is slightly less optimized, but suitable for a baseline JIT
compiler.

(Copy-and-patch JIT compiler contributed by Brandt Bucher,
directly inspired by the paper
`Copy-and-Patch Compilation <https://fredrikbk.com/publications/copy-and-patch.pdf>`_
by Haoran Xu and Fredrik Kjolstad. For more information,
`a talk <https://youtu.be/HxSHIpEQRjs?si=RwC78FcXrThIgFmY>`_ is available.)

Results and Future Work
-----------------------

> The final performance results will be updated before CPython 3.13's release.

Reviewer 2: Similarly:
```suggestion
The final performance results will be updated before CPython 3.13's beta release.
```

programs. The internal architecture is roughly as follows.

Intermediate Representation
---------------------------

We start with specialized *Tier 1 bytecode*.
See :ref:`What's new in 3.11 <whatsnew311-pep659>` for details.

When the Tier 1 bytecode gets hot enough, the interpreter creates
straight-line sequences of bytecode known as "traces", and translates that
to a new, purely internal *Tier 2 IR*, a.k.a. micro-ops ("uops").
These straight-line sequences can cross function call boundaries,
allowing more effective optimizations, listed in the next section.

The Tier 2 IR uses the same stack-based VM as Tier 1, but the
instruction format is better suited to translation to machine code.

(Tier 2 IR contributed by Mark Shannon and Guido van Rossum.)

Optimizations
-------------

We have several optimization and analysis passes for Tier 2 IR, which
are applied before Tier 2 IR is interpreted or translated to machine code.
These optimizations take unoptimized Tier 2 IR and produce optimized Tier 2
IR:

* Type propagation -- through forward data-flow analysis, we infer
  and deduce information about types. This allows us to eliminate

>   much of the overhead associated with dynamic typing in the future.

Reviewer 4: Could use explanation of "forward data-flow analysis"? The second sentence is highly speculative, I would express a more short-term and concrete benefit.

* Constant propagation -- through forward data-flow analysis, we can reduce

>   expressions like ::

Reviewer 4: The example does not show expressions but statements.
```suggestion
  statements like ::
```

    a = 1
    b = 2
    c = a + b

 to ::

    a = 1
    b = 2
    c = 3

* Guard elimination -- through a combination of constant and type information,
  we can eliminate type checks and other guards associated with operations.
  These guards validate specialized operations, but add a slight bit of
  overhead. For example, integer addition needs a type check that checks
  both operands are integers. As a proof of concept, we managed to eliminate

>   over 70% of integer type checks in our own benchmarks.

Reviewer 4: This last sentence reads more like an academic report about your work than as an explanation.

* Loop splitting -- after the first iteration, we gain a lot more type
  information. Thus, we peel the first iteration of loops to produce
  an optimized body that exploits this additional type information.
  This also achieves a similar effect to an optimization called
  loop-invariant code motion, but only for guards.

* Globals to constant promotion -- global value loads become constant

>   loads, speeding them up and also allowing for more constant propagation.

Reviewer 4: IIUC this work is independent from the data-flow analysis part. It relies on dictionary watchers, which deserve a separate mention if they aren't already mentioned.

Author: dictionary watchers were implemented in 3.12, not 3.13. So it's already mentioned there.

* This section is non-exhaustive and will be updated with further

>   optimizations, until CPython 3.13's beta release.

Reviewer 4: Maybe this should be at the top.

(Tier 2 optimizer contributed by Ken Jin, with implementation help
by Guido van Rossum, Mark Shannon, and Jules Poon. Special thanks

> to Manuel Rigger and Martin Henz.)

Reviewer 4: I think we have to be a bit more careful here giving credit where due.

Author: I will remove Jules and Martin but I will insist on crediting Manuel, I've met with him consistently every 2 weeks to discuss this work (and adjacent experiments on optimizing CPython), and its technical details. This has been ongoing for over 5 months now. So I'm very thankful for his help.

Execution Engine
----------------

There are two execution engines for Tier 2 IR:
the Tier 2 interpreter and the Just-in-Time (JIT) compiler.

The Tier 2 interpreter is mostly intended for debugging
the earlier stages of the optimization pipeline. If the JIT is not
enabled, the Tier 2 interpreter can be invoked by passing Python the
``-X uops`` option or by setting the ``PYTHON_UOPS`` environment
variable to ``1``.

The second is the JIT compiler. When the ``--enable-experimental-jit``
build-time option is used, the optimized Tier 2 IR is translated to machine
code, which is then executed. This does not require additional
runtime options.

The machine code translation process uses a technique called
*copy-and-patch*. It has no runtime dependencies, but there is a new
build-time dependency on `LLVM <https://llvm.org>`_.
The main benefit of this technique is
fast compilation, reported as orders of magnitudes faster versus

> traditional compilation techniques in the paper linked below. The code

Reviewer 4: Perhaps add an explanation why fast compilation is important.

produced is slightly less optimized, but suitable for a baseline JIT
compiler.

(Copy-and-patch JIT compiler contributed by Brandt Bucher,
directly inspired by the paper
`Copy-and-Patch Compilation <https://fredrikbk.com/publications/copy-and-patch.pdf>`_
by Haoran Xu and Fredrik Kjolstad. For more information,

> `a talk <https://youtu.be/HxSHIpEQRjs?si=RwC78FcXrThIgFmY>`_ is available.)

Reviewer 4: Please name the author of the talk. (Aren't there also other info resources about the JIT compiler by @brandtbucher?)

Results and Future Work
-----------------------

> The final performance results will be updated before CPython 3.13's beta release.

Reviewer 4: Add "here"?

The JIT compiler is rather unoptimized, and serves as the foundation
for significant optimizations in future releases. As such, we do not
expect the first iteration of the JIT compiler to produce a significant

> speedup.

Reviewer 4: Not sure we need to drag it down that much.

About
-----

This work was done by the Faster CPython team, and many other external
contributors. The team consists of engineers from Microsoft, Meta,
Quansight, and Bloomberg, who are either paid in part to do this, or

> volunteer in their free time.

Reviewer 4: I don't see the point of this section -- contributors are already listed above. If and when Microsoft wants its name in the credits I will let you know, but not now.

Author: We had a similar credits section in 3.11, so this is mirroring that, but I will remove it. https://docs.python.org/3/whatsnew/3.11.html#about

Reviewer 4: > We had a similar credits section in 3.11, so this is mirroring that, but I will remove it. https://docs.python.org/3/whatsnew/3.11.html#about

Thanks. It's tradition to credit individuals regardless of who paid for their time.

Intermediate Representation
---------------------------

We start with specialized *Tier 1 bytecode*.
See :ref:`What's new in 3.11 <whatsnew311-pep659>` for details.

When the Tier 1 bytecode gets hot enough, the interpreter creates
straight-line sequences of bytecode known as "traces", and translates that
to a new, purely internal *Tier 2 IR*, a.k.a. micro-ops ("uops").
These straight-line sequences can cross function call boundaries,
allowing more effective optimizations, listed in the next section.

The Tier 2 IR uses the same stack-based VM as Tier 1, but the
instruction format is better suited to translation to machine code.

(Tier 2 IR contributed by Mark Shannon and Guido van Rossum.)

Optimizations
-------------

We have several optimization and analysis passes for Tier 2 IR, which
are applied before Tier 2 IR is interpreted or translated to machine code.
These optimizations take unoptimized Tier 2 IR and produce optimized Tier 2
IR:

* Type propagation -- through forward data-flow analysis, we infer
  and deduce information about types. This allows us to eliminate
  much of the overhead associated with dynamic typing in the future.

> * Constant propagation -- through forward data-flow analysis, we can reduce

Reviewer 5: General comment: I wonder if for people who are compiler noobs but like to learn more it would be nice to link to wikipedia or other accessible sources? In this case, it is easy to google and find this: https://en.wikipedia.org/wiki/Constant_folding but in other cases it might not be?

(to be clear, I love the current text, it is very clearly written :heart:, so this does not seem as important)

Author: Thanks for the kind words! I think most of the terms are explained (I try to give a short summary of what they mean in each bullet point). Not all of this can be found in Wikipedia articles though, and some are found in particular programming language research papers.

  expressions like ::

    a = 1
    b = 2
    c = a + b

 to ::

    a = 1
    b = 2
    c = 3

> * Guard elimination -- through a combination of constant and type information,

Reviewer 5: I wonder if the guard concept could be expanded a bit more. The example of "type checking" makes things fairly clear what it means, but wonder if it could be expanded.

  of integer type checks in our own benchmarks.

* Loop splitting -- after the first iteration, we gain a lot more type
  information. Thus, we peel the first iteration of loops to produce
  an optimized body that exploits this additional type information.
  This also achieves a similar effect to an optimization called
  loop-invariant code motion, but only for guards.

* Globals to constant promotion -- global value loads become constant
  loads, speeding them up and also allowing for more constant propagation.

* This section is non-exhaustive and will be updated with further
  optimizations, up till CPython 3.13's release.

(Tier 2 optimizer contributed by Ken Jin, with implementation help
by Guido van Rossum, Mark Shannon, and Jules Poon. Special thanks
to Manuel Rigger and Martin Henz.)

Execution Engine
----------------

There are two execution engines for Tier 2 IR.

The first is the Tier 2 interpreter, but it is mostly intended for debugging
the earlier stages of the optimization pipeline. If the JIT is not
enabled, the Tier 2 interpreter can be invoked by passing Python the
``-X uops`` option or by setting the ``PYTHON_UOPS`` environment
variable to ``1``.

> The second is machine code. When the ``--enable-experimental-jit``

Reviewer 5: Maybe to clarify

```suggestion
The second is machine code. When the ``--enable-experimental-jit`` build-time
```

option is used, the optimized Tier 2 IR is translated to machine
code, which is then executed. This does not require additional
runtime options.

The machine code translation process uses an architecture called
*copy-and-patch*. It has no runtime dependencies, but there is a new
build-time dependency on LLVM.

(Copy-and-patch JIT compiler contributed by Brandt Bucher,

> directly inspired by the paper "Copy-and-Patch Compilation"

Reviewer 5: Add a link to the paper, and perhaps other more accessible resources (blog posts, Brandt's talk...)?
